### PR TITLE
Fixing drawContours() function to correctly draw child contours

### PR DESCRIPTION
**Pull request by shyama7004**
### Steps Followed

1. **Recursive Drawing**: Implemented a recursive function to draw child contours, ensuring no redundant fillings occur.
2. **Thickness Issue Resolved**: Fixed the inability to draw child contours when `thickness >= 0`.

#### Previously, when `thickness = 2`:

Previous Contour Error

#### After the Fix, when `thickness = 2`:

Positive Thick

#### After the Fix, when `thickness = 0`:

thickness = 0

#### After the Fix, when `thickness = -2`:

thickness =  -2

### Other Details

- **Tested on**: macOS
- **OpenCV Version**: 4.10.0
- **Link to the Issue**: [Issue #26264](https://github.com/opencv/opencv/issues/26264)

Please review the code and provide any suggestions for improvements or feedback.

## Checklist

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by shyama7004** [comment]
I'm concerned that the current approach might introduce overhead errors. After working on the iterative method for 11 hours, I'm considering implementing it. Could you provide feedback on whether this would be a good approach?(It's giving the same results)

### here's the enhanced code

```cpp
namespace cv
{
    void drawContourIteratively(Mat& image, InputArrayOfArrays _contours, int idx, const Scalar& color, int thickness, int lineType, InputArray _hierarchy, Point offset)
    {
        Mat hierarchy = _hierarchy.getMat();
        double color_buf[4]; // this holds the raw data of the color
        scalarToRawData(color, color_buf, image.type(), 0); // converts cv::Scalar to raw data

        std::stack<int> stack;
        stack.push(idx);

        while (!stack.empty())
        {
            int currentIdx = stack.top();
            stack.pop();

            Mat cnt = _contours.getMat(currentIdx);
            if (!cnt.empty())
            {
                const int npoints = cnt.checkVector(2, CV_32S);
                CV_Assert(npoints > 0);
                for (int j = 0; j < npoints; ++j)
                {
                    const bool isLastIter = j == npoints - 1;
                    const Point pt1 = cnt.at<Point>(j);
                    const Point pt2 = cnt.at<Point>(isLastIter ? 0 : j + 1);

                    ThickLine(image, pt1 + offset, pt2 + offset, color_buf, thickness, lineType, 2, 0);
                }
            }

            // pushes all child contours onto the stack
            int childIdx = hierarchy.at<Vec4i>(currentIdx)[2]; // first child
            while (childIdx >= 0)
            {
                stack.push(childIdx);
                childIdx = hierarchy.at<Vec4i>(childIdx)[0]; // next sibling
            }
        }
    }
}


void cv::drawContours(InputOutputArray _image, InputArrayOfArrays _contours,
                      int contourIdx, const Scalar& color, int thickness,
                      int lineType, InputArray _hierarchy,
                      int maxLevel, Point offset)
{
    CV_INSTRUMENT_REGION();
    CV_Assert(thickness <= MAX_THICKNESS);
    const size_t ncontours = _contours.total();
    if (!ncontours)
        return;

    CV_Assert(ncontours <= (size_t)std::numeric_limits<int>::max());
    if (lineType == cv::LINE_AA && _image.depth() != CV_8U)
        lineType = 8;
    Mat image = _image.getMat(), hierarchy = _hierarchy.getMat();

    if (thickness >= 0) // contour lines
    {
        double color_buf[4] {};
        scalarToRawData(color, color_buf, _image.type(), 0);


        if (contourIdx >= 0)
        {
            drawContourIteratively(image, _contours, contourIdx, color, thickness, lineType, _hierarchy, offset);
        }
        else
        {
            // iterates over all contours and draws them and their childrens
            for (int i = 0; i < (int)ncontours; ++i)
            {
                drawContourIteratively(image, _contours, i, color, thickness, lineType, _hierarchy, offset);
            }
        }
    }
    else // filled polygons
    {
        int i = 0, end = (int)ncontours;
        if (contourIdx >= 0)
        {
            i = contourIdx;
            end = i + 1;
        }
        std::vector<int> indexesToFill;
        if (hierarchy.empty() || maxLevel == 0)
        {
            for (; i != end; ++i)
                indexesToFill.push_back(i);
        }
        else
        {
            std::stack<int> indexes;
            for (; i != end; ++i)
            {
                // either all from the top level or a single contour
                if (hierarchy.at<cv::Vec4i>(i)[3] < 0 || contourIdx >= 0)
                    indexes.push(i);
            }
            while (!indexes.empty())
            {
                // to get the current element
                const int cur = indexes.top();
                indexes.pop();

                //  checks the current element depth
                int curLevel = -1;
                int par = cur;
                while (par >= 0)
                {
                    par = hierarchy.at<cv::Vec4i>(par)[3]; // parent
                    ++curLevel;
                }
                if (curLevel <= maxLevel)
                {
                    indexesToFill.push_back(cur);
                }

                int next = hierarchy.at<cv::Vec4i>(cur)[2]; // first child
                while (next > 0)
                {
                    indexes.push(next);
                    next = hierarchy.at<cv::Vec4i>(next)[0]; // next sibling
                }
            }
        }
        std::vector<cv::Mat> contoursToFill;
        for (const int &idx : indexesToFill)
            contoursToFill.push_back(_contours.getMat(idx));
        fillPoly(image, contoursToFill, color, lineType, 0, offset);
    }
}
```

**Comment by asmorkalov** [comment]
cc @mshabunin

File: modules/imgproc/src/drawing.cpp [hunk 1]
```diff
@@ -2514,54 +2539,71 @@ void cv::drawContours( InputOutputArray _image, InputArrayOfArrays _contours,
...
         std::vector<Mat> contoursToFill;
-        for (const int & idx : indexesToFill)
+        for (const int &idx : indexesToFill)
             contoursToFill.push_back(_contours.getMat(idx));
         fillPoly(image, contoursToFill, color, lineType, 0, offset);
     }
 }
 
+// this particular function recursively draws each contour
+void cv::drawContourRecursively(Mat &image, InputArrayOfArrays _contours, int idx, const Scalar &color, int thickness, int lineType, InputArray _hierarchy, Point offset)
```
Comment (mshabunin) [inline modules/imgproc/src/drawing.cpp:2586]: This function has a misleading name - it is not recursive.

File: modules/imgproc/src/drawing.cpp [hunk 2]
```diff
@@ -2514,54 +2539,71 @@ void cv::drawContours( InputOutputArray _image, InputArrayOfArrays _contours,
...
+            while (next > 0)
+            {
+                indexes.push(next);
+                next = hierarchy.at<Vec4i>(next)[0]; // next sibling
             }
         }
+
         std::vector<Mat> contoursToFill;
-        for (const int & idx : indexesToFill)
+        for (const int &idx : indexesToFill)
```
Comment (mshabunin) [inline modules/imgproc/src/drawing.cpp:2579]: Please avoid unrelated changes.

**Review by mshabunin — COMMENTED** [review]
I believe we have to start with a test reproducing the issue.

File: modules/imgproc/src/drawing.cpp (hunk 2 above)
Comment (shyama7004) [reply modules/imgproc/src/drawing.cpp:2579]: Ok i willl

**Comment by vpisarev** [comment]
@shyama7004, thank you. It will take us some time to analyze the problem and your solution, let's do it after upcoming OpenCV 5.0 alpha (this month). One thing that would definitely be useful to add, as @mshabunin says, is to add some regression test that would check the original problem and the solution.

and I also agree that whitespace modifications should be minimized, to keep the patch as compact and easy to observe as possible.

**Comment by shyama7004** [comment]
Thank you for your feedback, @vpisarev. I am currently addressing the issues and working on potential solutions.

**Comment by sturkmen72** [comment]
@shyama7004 as a contributor i can give some advices about contribution ideas. you can find my e-mail on github. lets discuss.

**Comment by shyama7004** [comment]
Thank you , 
I will mail shortly @sturkmen72.
